lucene_qat_wrapper/lucene_with_qat.diff

```diff
@@ -0,0 +1,588 @@
+@@ -43,6 +46,7 @@
+     <pathelement location="${common.dir}/build/codecs/classes/java"/>
+     <pathelement location="${common.dir}/build/test-framework/classes/java"/>
+     <path refid="junit-path"/>
++    <path refid="classpath"/>
+     <pathelement location="${-test.classpath.java9.addon}"/><!-- if it's a duplicate it gets removed by Ant! -->
+     <pathelement location="${build.dir}/classes/java"/>
+     <pathelement location="${build.dir}/classes/test"/>
+ 
+diff --git a/lucene/core/src/java/org/apache/lucene/codecs/compressing/CompressionMode.java b/lucene/core/src/java/org/apache/lucene/codecs/compressing/CompressionMode.java
+index 53a84cb..3af4b6b 100644
+--- a/lucene/core/src/java/org/apache/lucene/codecs/compressing/CompressionMode.java
++++ b/lucene/core/src/java/org/apache/lucene/codecs/compressing/CompressionMode.java
+@@ -22,6 +22,8 @@ import java.util.zip.DataFormatException;
+ import java.util.zip.Deflater;
+ import java.util.zip.Inflater;
+ 
++import com.intel.qat.jni.QatCompressorJNI;
++import com.intel.qat.jni.QatDecompressorJNI;
+ import org.apache.lucene.index.CorruptIndexException;
+ import org.apache.lucene.store.DataInput;
+ import org.apache.lucene.store.DataOutput;
+@@ -113,8 +114,30 @@ public abstract class CompressionMode {
+ 
+   };
+ 
+-  /** Sole constructor. */
+-  protected CompressionMode() {}
++  /**
++   * This compression mode is using the QAT
++   */
++  public static final CompressionMode QAT = new CompressionMode() {
++
++    @Override
++    public Compressor newCompressor() {
++      return new QatCompressor();
++    }
++
++    @Override
++    public Decompressor newDecompressor() {
++      return new QatDecompressor();
++    }
++
++    @Override
++    public String toString() {
++      return "QAT";
++    }
++  };
++
++  /**Sole constructor.*/
++  protected CompressionMode() {
++  }
+ 
+   /**
+    * Create a new {@link Compressor} instance.
+@@ -296,4 +319,111 @@ public abstract class CompressionMode {
+ 
+   }
+ 
++  private static final class QatDecompressor extends Decompressor {
++
++    byte[] compressed;
++    int directBufferSize = 655360;
++
++    QatDecompressor() {
++      compressed = new byte[0];
++    }
++
++    QatDecompressor(int directBufferSize) {
++      this.directBufferSize = directBufferSize;
++      compressed = new byte[0];
++    }
++
++    @Override
++    public void decompress(DataInput in, int originalLength, int offset, int length, BytesRef bytes) throws IOException {
++      assert offset + length <= originalLength;
++      if (length == 0) {
++        bytes.length = 0;
++        return;
++      }
++      final int compressedLength = in.readVInt();
++      // pad with extra "dummy byte": see javadocs for using Inflater(true)
++      // we do it for compliance, but it's unnecessary for years in zlib.
++      final int paddedLength = compressedLength;
++      compressed = ArrayUtil.grow(compressed, paddedLength + 1);
```

Comment:
Resizing is very expensive operation for array. What's the main purpose to hold the compressed as a local variable? We can just create it on the fly.